{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Helper Functions For Preprocessing Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from helper_functions import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Preprocessing"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_position</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>261</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>316</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_id    transcript_id  transcript_position  label\n",
       "0  ENSG00000004059  ENST00000000233                  244      0\n",
       "1  ENSG00000004059  ENST00000000233                  261      0\n",
       "2  ENSG00000004059  ENST00000000233                  316      0"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "labels = pd.read_csv('../data/data.info.labelled')\n",
    "labels.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>position</th>\n",
       "      <th>sequence</th>\n",
       "      <th>reads</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>[[0.00299, 2.06, 125.0, 0.0177, 10.4, 122.0, 0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>261</td>\n",
       "      <td>CAAACTG</td>\n",
       "      <td>[[0.0126, 1.95, 111.0, 0.0125, 1.27, 108.0, 0....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>316</td>\n",
       "      <td>GAAACAG</td>\n",
       "      <td>[[0.00432, 2.02, 104.0, 0.00299, 3.56, 99.3, 0...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     transcript_id  position sequence  \\\n",
       "0  ENST00000000233       244  AAGACCA   \n",
       "1  ENST00000000233       261  CAAACTG   \n",
       "2  ENST00000000233       316  GAAACAG   \n",
       "\n",
       "                                               reads  \n",
       "0  [[0.00299, 2.06, 125.0, 0.0177, 10.4, 122.0, 0...  \n",
       "1  [[0.0126, 1.95, 111.0, 0.0125, 1.27, 108.0, 0....  \n",
       "2  [[0.00432, 2.02, 104.0, 0.00299, 3.56, 99.3, 0...  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = load_data_to_dataframe('../data/dataset0.json.gz')\n",
    "df.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Assign labels to the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>position</th>\n",
       "      <th>sequence</th>\n",
       "      <th>reads</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>[[0.00299, 2.06, 125.0, 0.0177, 10.4, 122.0, 0...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>261</td>\n",
       "      <td>CAAACTG</td>\n",
       "      <td>[[0.0126, 1.95, 111.0, 0.0125, 1.27, 108.0, 0....</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>316</td>\n",
       "      <td>GAAACAG</td>\n",
       "      <td>[[0.00432, 2.02, 104.0, 0.00299, 3.56, 99.3, 0...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_id    transcript_id  position sequence  \\\n",
       "0  ENSG00000004059  ENST00000000233       244  AAGACCA   \n",
       "1  ENSG00000004059  ENST00000000233       261  CAAACTG   \n",
       "2  ENSG00000004059  ENST00000000233       316  GAAACAG   \n",
       "\n",
       "                                               reads  label  \n",
       "0  [[0.00299, 2.06, 125.0, 0.0177, 10.4, 122.0, 0...      0  \n",
       "1  [[0.0126, 1.95, 111.0, 0.0125, 1.27, 108.0, 0....      0  \n",
       "2  [[0.00432, 2.02, 104.0, 0.00299, 3.56, 99.3, 0...      0  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = combine_data(df, labels)\n",
    "df.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Extract mean reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>position</th>\n",
       "      <th>sequence</th>\n",
       "      <th>reads</th>\n",
       "      <th>label</th>\n",
       "      <th>mean_reads</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>[[0.00299, 2.06, 125.0, 0.0177, 10.4, 122.0, 0...</td>\n",
       "      <td>0</td>\n",
       "      <td>[0.008264378378378385, 4.223783783783786, 123....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>261</td>\n",
       "      <td>CAAACTG</td>\n",
       "      <td>[[0.0126, 1.95, 111.0, 0.0125, 1.27, 108.0, 0....</td>\n",
       "      <td>0</td>\n",
       "      <td>[0.006609244186046515, 3.2164244186046504, 109...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>316</td>\n",
       "      <td>GAAACAG</td>\n",
       "      <td>[[0.00432, 2.02, 104.0, 0.00299, 3.56, 99.3, 0...</td>\n",
       "      <td>0</td>\n",
       "      <td>[0.0075699999999999995, 2.94054054054054, 105....</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_id    transcript_id  position sequence  \\\n",
       "0  ENSG00000004059  ENST00000000233       244  AAGACCA   \n",
       "1  ENSG00000004059  ENST00000000233       261  CAAACTG   \n",
       "2  ENSG00000004059  ENST00000000233       316  GAAACAG   \n",
       "\n",
       "                                               reads  label  \\\n",
       "0  [[0.00299, 2.06, 125.0, 0.0177, 10.4, 122.0, 0...      0   \n",
       "1  [[0.0126, 1.95, 111.0, 0.0125, 1.27, 108.0, 0....      0   \n",
       "2  [[0.00432, 2.02, 104.0, 0.00299, 3.56, 99.3, 0...      0   \n",
       "\n",
       "                                          mean_reads  \n",
       "0  [0.008264378378378385, 4.223783783783786, 123....  \n",
       "1  [0.006609244186046515, 3.2164244186046504, 109...  \n",
       "2  [0.0075699999999999995, 2.94054054054054, 105....  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = extract_mean_reads(df)\n",
    "df.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Train test split by gene_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>position</th>\n",
       "      <th>sequence</th>\n",
       "      <th>reads</th>\n",
       "      <th>label</th>\n",
       "      <th>mean_reads</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ENSG00000003056</td>\n",
       "      <td>ENST00000000412</td>\n",
       "      <td>355</td>\n",
       "      <td>GAAACTA</td>\n",
       "      <td>[[0.00232, 2.41, 109.0, 0.0222, 2.85, 111.0, 0...</td>\n",
       "      <td>0</td>\n",
       "      <td>[0.007340399999999998, 2.9771799999999997, 108...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ENSG00000003056</td>\n",
       "      <td>ENST00000000412</td>\n",
       "      <td>367</td>\n",
       "      <td>GGGACCG</td>\n",
       "      <td>[[0.00232, 1.32, 117.0, 0.0073, 7.89, 120.0, 0...</td>\n",
       "      <td>0</td>\n",
       "      <td>[0.00898787234042553, 3.961489361702128, 118.6...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ENSG00000003056</td>\n",
       "      <td>ENST00000000412</td>\n",
       "      <td>496</td>\n",
       "      <td>AGGACTG</td>\n",
       "      <td>[[0.00398, 2.46, 111.0, 0.016, 3.36, 125.0, 0....</td>\n",
       "      <td>0</td>\n",
       "      <td>[0.011064705882352935, 7.299607843137254, 115....</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            gene_id    transcript_id  position sequence  \\\n",
       "18  ENSG00000003056  ENST00000000412       355  GAAACTA   \n",
       "19  ENSG00000003056  ENST00000000412       367  GGGACCG   \n",
       "20  ENSG00000003056  ENST00000000412       496  AGGACTG   \n",
       "\n",
       "                                                reads  label  \\\n",
       "18  [[0.00232, 2.41, 109.0, 0.0222, 2.85, 111.0, 0...      0   \n",
       "19  [[0.00232, 1.32, 117.0, 0.0073, 7.89, 120.0, 0...      0   \n",
       "20  [[0.00398, 2.46, 111.0, 0.016, 3.36, 125.0, 0....      0   \n",
       "\n",
       "                                           mean_reads  \n",
       "18  [0.007340399999999998, 2.9771799999999997, 108...  \n",
       "19  [0.00898787234042553, 3.961489361702128, 118.6...  \n",
       "20  [0.011064705882352935, 7.299607843137254, 115....  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "(97073, 7)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>position</th>\n",
       "      <th>sequence</th>\n",
       "      <th>reads</th>\n",
       "      <th>label</th>\n",
       "      <th>mean_reads</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>[[0.00299, 2.06, 125.0, 0.0177, 10.4, 122.0, 0...</td>\n",
       "      <td>0</td>\n",
       "      <td>[0.008264378378378385, 4.223783783783786, 123....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>261</td>\n",
       "      <td>CAAACTG</td>\n",
       "      <td>[[0.0126, 1.95, 111.0, 0.0125, 1.27, 108.0, 0....</td>\n",
       "      <td>0</td>\n",
       "      <td>[0.006609244186046515, 3.2164244186046504, 109...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>316</td>\n",
       "      <td>GAAACAG</td>\n",
       "      <td>[[0.00432, 2.02, 104.0, 0.00299, 3.56, 99.3, 0...</td>\n",
       "      <td>0</td>\n",
       "      <td>[0.0075699999999999995, 2.94054054054054, 105....</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_id    transcript_id  position sequence  \\\n",
       "0  ENSG00000004059  ENST00000000233       244  AAGACCA   \n",
       "1  ENSG00000004059  ENST00000000233       261  CAAACTG   \n",
       "2  ENSG00000004059  ENST00000000233       316  GAAACAG   \n",
       "\n",
       "                                               reads  label  \\\n",
       "0  [[0.00299, 2.06, 125.0, 0.0177, 10.4, 122.0, 0...      0   \n",
       "1  [[0.0126, 1.95, 111.0, 0.0125, 1.27, 108.0, 0....      0   \n",
       "2  [[0.00432, 2.02, 104.0, 0.00299, 3.56, 99.3, 0...      0   \n",
       "\n",
       "                                          mean_reads  \n",
       "0  [0.008264378378378385, 4.223783783783786, 123....  \n",
       "1  [0.006609244186046515, 3.2164244186046504, 109...  \n",
       "2  [0.0075699999999999995, 2.94054054054054, 105....  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "(24765, 7)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Train test split by gene id\n",
    "from sklearn.model_selection import train_test_split\n",
    "train_gene_ids, test_gene_ids = train_test_split(df['gene_id'].unique(), test_size=0.2, random_state=4262)\n",
    "\n",
    "train_df = df[df['gene_id'].isin(train_gene_ids)].copy()\n",
    "test_df = df[df['gene_id'].isin(test_gene_ids)].copy()\n",
    "\n",
    "display(train_df.head(3), train_df.shape)\n",
    "display(test_df.head(3), test_df.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Scale mean reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Scaler saved to ../artifacts/mean_reads_scaler.pkl\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>position</th>\n",
       "      <th>sequence</th>\n",
       "      <th>reads</th>\n",
       "      <th>label</th>\n",
       "      <th>mean_reads</th>\n",
       "      <th>scaled_mean_reads</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ENSG00000003056</td>\n",
       "      <td>ENST00000000412</td>\n",
       "      <td>355</td>\n",
       "      <td>GAAACTA</td>\n",
       "      <td>[[0.00232, 2.41, 109.0, 0.0222, 2.85, 111.0, 0...</td>\n",
       "      <td>0</td>\n",
       "      <td>[0.007340399999999998, 2.9771799999999997, 108...</td>\n",
       "      <td>[-0.43079389735504386, -0.7090163040809951, -0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ENSG00000003056</td>\n",
       "      <td>ENST00000000412</td>\n",
       "      <td>367</td>\n",
       "      <td>GGGACCG</td>\n",
       "      <td>[[0.00232, 1.32, 117.0, 0.0073, 7.89, 120.0, 0...</td>\n",
       "      <td>0</td>\n",
       "      <td>[0.00898787234042553, 3.961489361702128, 118.6...</td>\n",
       "      <td>[0.49106279705280276, -0.1994143266757514, 0.6...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ENSG00000003056</td>\n",
       "      <td>ENST00000000412</td>\n",
       "      <td>496</td>\n",
       "      <td>AGGACTG</td>\n",
       "      <td>[[0.00398, 2.46, 111.0, 0.016, 3.36, 125.0, 0....</td>\n",
       "      <td>0</td>\n",
       "      <td>[0.011064705882352935, 7.299607843137254, 115....</td>\n",
       "      <td>[1.6531720821786586, 1.5288144651450055, 0.418...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            gene_id    transcript_id  position sequence  \\\n",
       "18  ENSG00000003056  ENST00000000412       355  GAAACTA   \n",
       "19  ENSG00000003056  ENST00000000412       367  GGGACCG   \n",
       "20  ENSG00000003056  ENST00000000412       496  AGGACTG   \n",
       "\n",
       "                                                reads  label  \\\n",
       "18  [[0.00232, 2.41, 109.0, 0.0222, 2.85, 111.0, 0...      0   \n",
       "19  [[0.00232, 1.32, 117.0, 0.0073, 7.89, 120.0, 0...      0   \n",
       "20  [[0.00398, 2.46, 111.0, 0.016, 3.36, 125.0, 0....      0   \n",
       "\n",
       "                                           mean_reads  \\\n",
       "18  [0.007340399999999998, 2.9771799999999997, 108...   \n",
       "19  [0.00898787234042553, 3.961489361702128, 118.6...   \n",
       "20  [0.011064705882352935, 7.299607843137254, 115....   \n",
       "\n",
       "                                    scaled_mean_reads  \n",
       "18  [-0.43079389735504386, -0.7090163040809951, -0...  \n",
       "19  [0.49106279705280276, -0.1994143266757514, 0.6...  \n",
       "20  [1.6531720821786586, 1.5288144651450055, 0.418...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>position</th>\n",
       "      <th>sequence</th>\n",
       "      <th>reads</th>\n",
       "      <th>label</th>\n",
       "      <th>mean_reads</th>\n",
       "      <th>scaled_mean_reads</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>[[0.00299, 2.06, 125.0, 0.0177, 10.4, 122.0, 0...</td>\n",
       "      <td>0</td>\n",
       "      <td>[0.008264378378378385, 4.223783783783786, 123....</td>\n",
       "      <td>[0.08622580166837883, -0.06361783691765782, 1....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>261</td>\n",
       "      <td>CAAACTG</td>\n",
       "      <td>[[0.0126, 1.95, 111.0, 0.0125, 1.27, 108.0, 0....</td>\n",
       "      <td>0</td>\n",
       "      <td>[0.006609244186046515, 3.2164244186046504, 109...</td>\n",
       "      <td>[-0.8399181449731496, -0.5851533870472007, -0....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>316</td>\n",
       "      <td>GAAACAG</td>\n",
       "      <td>[[0.00432, 2.02, 104.0, 0.00299, 3.56, 99.3, 0...</td>\n",
       "      <td>0</td>\n",
       "      <td>[0.0075699999999999995, 2.94054054054054, 105....</td>\n",
       "      <td>[-0.30231932947163226, -0.7279854836073506, -0...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_id    transcript_id  position sequence  \\\n",
       "0  ENSG00000004059  ENST00000000233       244  AAGACCA   \n",
       "1  ENSG00000004059  ENST00000000233       261  CAAACTG   \n",
       "2  ENSG00000004059  ENST00000000233       316  GAAACAG   \n",
       "\n",
       "                                               reads  label  \\\n",
       "0  [[0.00299, 2.06, 125.0, 0.0177, 10.4, 122.0, 0...      0   \n",
       "1  [[0.0126, 1.95, 111.0, 0.0125, 1.27, 108.0, 0....      0   \n",
       "2  [[0.00432, 2.02, 104.0, 0.00299, 3.56, 99.3, 0...      0   \n",
       "\n",
       "                                          mean_reads  \\\n",
       "0  [0.008264378378378385, 4.223783783783786, 123....   \n",
       "1  [0.006609244186046515, 3.2164244186046504, 109...   \n",
       "2  [0.0075699999999999995, 2.94054054054054, 105....   \n",
       "\n",
       "                                   scaled_mean_reads  \n",
       "0  [0.08622580166837883, -0.06361783691765782, 1....  \n",
       "1  [-0.8399181449731496, -0.5851533870472007, -0....  \n",
       "2  [-0.30231932947163226, -0.7279854836073506, -0...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Scale mean reads of train data first, then use the fitted scaler to scale test data\n",
    "scaler, scaled_train_df = scale_mean_reads(train_df)\n",
    "scaler, scaled_test_df = scale_mean_reads(test_df, scaler=scaler)\n",
    "\n",
    "display(scaled_train_df.head(3))\n",
    "display(scaled_test_df.head(3))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Extract middle sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>position</th>\n",
       "      <th>sequence</th>\n",
       "      <th>reads</th>\n",
       "      <th>label</th>\n",
       "      <th>mean_reads</th>\n",
       "      <th>scaled_mean_reads</th>\n",
       "      <th>middle_sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ENSG00000003056</td>\n",
       "      <td>ENST00000000412</td>\n",
       "      <td>355</td>\n",
       "      <td>GAAACTA</td>\n",
       "      <td>[[0.00232, 2.41, 109.0, 0.0222, 2.85, 111.0, 0...</td>\n",
       "      <td>0</td>\n",
       "      <td>[0.007340399999999998, 2.9771799999999997, 108...</td>\n",
       "      <td>[-0.43079389735504386, -0.7090163040809951, -0...</td>\n",
       "      <td>AAACT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ENSG00000003056</td>\n",
       "      <td>ENST00000000412</td>\n",
       "      <td>367</td>\n",
       "      <td>GGGACCG</td>\n",
       "      <td>[[0.00232, 1.32, 117.0, 0.0073, 7.89, 120.0, 0...</td>\n",
       "      <td>0</td>\n",
       "      <td>[0.00898787234042553, 3.961489361702128, 118.6...</td>\n",
       "      <td>[0.49106279705280276, -0.1994143266757514, 0.6...</td>\n",
       "      <td>GGACC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ENSG00000003056</td>\n",
       "      <td>ENST00000000412</td>\n",
       "      <td>496</td>\n",
       "      <td>AGGACTG</td>\n",
       "      <td>[[0.00398, 2.46, 111.0, 0.016, 3.36, 125.0, 0....</td>\n",
       "      <td>0</td>\n",
       "      <td>[0.011064705882352935, 7.299607843137254, 115....</td>\n",
       "      <td>[1.6531720821786586, 1.5288144651450055, 0.418...</td>\n",
       "      <td>GGACT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            gene_id    transcript_id  position sequence  \\\n",
       "18  ENSG00000003056  ENST00000000412       355  GAAACTA   \n",
       "19  ENSG00000003056  ENST00000000412       367  GGGACCG   \n",
       "20  ENSG00000003056  ENST00000000412       496  AGGACTG   \n",
       "\n",
       "                                                reads  label  \\\n",
       "18  [[0.00232, 2.41, 109.0, 0.0222, 2.85, 111.0, 0...      0   \n",
       "19  [[0.00232, 1.32, 117.0, 0.0073, 7.89, 120.0, 0...      0   \n",
       "20  [[0.00398, 2.46, 111.0, 0.016, 3.36, 125.0, 0....      0   \n",
       "\n",
       "                                           mean_reads  \\\n",
       "18  [0.007340399999999998, 2.9771799999999997, 108...   \n",
       "19  [0.00898787234042553, 3.961489361702128, 118.6...   \n",
       "20  [0.011064705882352935, 7.299607843137254, 115....   \n",
       "\n",
       "                                    scaled_mean_reads middle_sequence  \n",
       "18  [-0.43079389735504386, -0.7090163040809951, -0...           AAACT  \n",
       "19  [0.49106279705280276, -0.1994143266757514, 0.6...           GGACC  \n",
       "20  [1.6531720821786586, 1.5288144651450055, 0.418...           GGACT  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>position</th>\n",
       "      <th>sequence</th>\n",
       "      <th>reads</th>\n",
       "      <th>label</th>\n",
       "      <th>mean_reads</th>\n",
       "      <th>scaled_mean_reads</th>\n",
       "      <th>middle_sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>[[0.00299, 2.06, 125.0, 0.0177, 10.4, 122.0, 0...</td>\n",
       "      <td>0</td>\n",
       "      <td>[0.008264378378378385, 4.223783783783786, 123....</td>\n",
       "      <td>[0.08622580166837883, -0.06361783691765782, 1....</td>\n",
       "      <td>AGACC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>261</td>\n",
       "      <td>CAAACTG</td>\n",
       "      <td>[[0.0126, 1.95, 111.0, 0.0125, 1.27, 108.0, 0....</td>\n",
       "      <td>0</td>\n",
       "      <td>[0.006609244186046515, 3.2164244186046504, 109...</td>\n",
       "      <td>[-0.8399181449731496, -0.5851533870472007, -0....</td>\n",
       "      <td>AAACT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>316</td>\n",
       "      <td>GAAACAG</td>\n",
       "      <td>[[0.00432, 2.02, 104.0, 0.00299, 3.56, 99.3, 0...</td>\n",
       "      <td>0</td>\n",
       "      <td>[0.0075699999999999995, 2.94054054054054, 105....</td>\n",
       "      <td>[-0.30231932947163226, -0.7279854836073506, -0...</td>\n",
       "      <td>AAACA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_id    transcript_id  position sequence  \\\n",
       "0  ENSG00000004059  ENST00000000233       244  AAGACCA   \n",
       "1  ENSG00000004059  ENST00000000233       261  CAAACTG   \n",
       "2  ENSG00000004059  ENST00000000233       316  GAAACAG   \n",
       "\n",
       "                                               reads  label  \\\n",
       "0  [[0.00299, 2.06, 125.0, 0.0177, 10.4, 122.0, 0...      0   \n",
       "1  [[0.0126, 1.95, 111.0, 0.0125, 1.27, 108.0, 0....      0   \n",
       "2  [[0.00432, 2.02, 104.0, 0.00299, 3.56, 99.3, 0...      0   \n",
       "\n",
       "                                          mean_reads  \\\n",
       "0  [0.008264378378378385, 4.223783783783786, 123....   \n",
       "1  [0.006609244186046515, 3.2164244186046504, 109...   \n",
       "2  [0.0075699999999999995, 2.94054054054054, 105....   \n",
       "\n",
       "                                   scaled_mean_reads middle_sequence  \n",
       "0  [0.08622580166837883, -0.06361783691765782, 1....           AGACC  \n",
       "1  [-0.8399181449731496, -0.5851533870472007, -0....           AAACT  \n",
       "2  [-0.30231932947163226, -0.7279854836073506, -0...           AAACA  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "scaled_train_df_w_middle_seq = extract_middle_sequence(scaled_train_df)\n",
    "scaled_test_df_w_middle_seq = extract_middle_sequence(scaled_test_df)\n",
    "\n",
    "display(scaled_train_df_w_middle_seq.head(3))\n",
    "display(scaled_test_df_w_middle_seq.head(3))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "One hot encode middle sequence (DRACH motif)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DRACH Encoder saved to ../artifacts/drach_encoder.pkl\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>position</th>\n",
       "      <th>sequence</th>\n",
       "      <th>reads</th>\n",
       "      <th>label</th>\n",
       "      <th>mean_reads</th>\n",
       "      <th>scaled_mean_reads</th>\n",
       "      <th>middle_sequence</th>\n",
       "      <th>middle_sequence_OHE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ENSG00000003056</td>\n",
       "      <td>ENST00000000412</td>\n",
       "      <td>355</td>\n",
       "      <td>GAAACTA</td>\n",
       "      <td>[[0.00232, 2.41, 109.0, 0.0222, 2.85, 111.0, 0...</td>\n",
       "      <td>0</td>\n",
       "      <td>[0.007340399999999998, 2.9771799999999997, 108...</td>\n",
       "      <td>[-0.43079389735504386, -0.7090163040809951, -0...</td>\n",
       "      <td>AAACT</td>\n",
       "      <td>[0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ENSG00000003056</td>\n",
       "      <td>ENST00000000412</td>\n",
       "      <td>367</td>\n",
       "      <td>GGGACCG</td>\n",
       "      <td>[[0.00232, 1.32, 117.0, 0.0073, 7.89, 120.0, 0...</td>\n",
       "      <td>0</td>\n",
       "      <td>[0.00898787234042553, 3.961489361702128, 118.6...</td>\n",
       "      <td>[0.49106279705280276, -0.1994143266757514, 0.6...</td>\n",
       "      <td>GGACC</td>\n",
       "      <td>[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ENSG00000003056</td>\n",
       "      <td>ENST00000000412</td>\n",
       "      <td>496</td>\n",
       "      <td>AGGACTG</td>\n",
       "      <td>[[0.00398, 2.46, 111.0, 0.016, 3.36, 125.0, 0....</td>\n",
       "      <td>0</td>\n",
       "      <td>[0.011064705882352935, 7.299607843137254, 115....</td>\n",
       "      <td>[1.6531720821786586, 1.5288144651450055, 0.418...</td>\n",
       "      <td>GGACT</td>\n",
       "      <td>[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            gene_id    transcript_id  position sequence  \\\n",
       "18  ENSG00000003056  ENST00000000412       355  GAAACTA   \n",
       "19  ENSG00000003056  ENST00000000412       367  GGGACCG   \n",
       "20  ENSG00000003056  ENST00000000412       496  AGGACTG   \n",
       "\n",
       "                                                reads  label  \\\n",
       "18  [[0.00232, 2.41, 109.0, 0.0222, 2.85, 111.0, 0...      0   \n",
       "19  [[0.00232, 1.32, 117.0, 0.0073, 7.89, 120.0, 0...      0   \n",
       "20  [[0.00398, 2.46, 111.0, 0.016, 3.36, 125.0, 0....      0   \n",
       "\n",
       "                                           mean_reads  \\\n",
       "18  [0.007340399999999998, 2.9771799999999997, 108...   \n",
       "19  [0.00898787234042553, 3.961489361702128, 118.6...   \n",
       "20  [0.011064705882352935, 7.299607843137254, 115....   \n",
       "\n",
       "                                    scaled_mean_reads middle_sequence  \\\n",
       "18  [-0.43079389735504386, -0.7090163040809951, -0...           AAACT   \n",
       "19  [0.49106279705280276, -0.1994143266757514, 0.6...           GGACC   \n",
       "20  [1.6531720821786586, 1.5288144651450055, 0.418...           GGACT   \n",
       "\n",
       "                                  middle_sequence_OHE  \n",
       "18  [0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...  \n",
       "19  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...  \n",
       "20  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>position</th>\n",
       "      <th>sequence</th>\n",
       "      <th>reads</th>\n",
       "      <th>label</th>\n",
       "      <th>mean_reads</th>\n",
       "      <th>scaled_mean_reads</th>\n",
       "      <th>middle_sequence</th>\n",
       "      <th>middle_sequence_OHE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>[[0.00299, 2.06, 125.0, 0.0177, 10.4, 122.0, 0...</td>\n",
       "      <td>0</td>\n",
       "      <td>[0.008264378378378385, 4.223783783783786, 123....</td>\n",
       "      <td>[0.08622580166837883, -0.06361783691765782, 1....</td>\n",
       "      <td>AGACC</td>\n",
       "      <td>[0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>261</td>\n",
       "      <td>CAAACTG</td>\n",
       "      <td>[[0.0126, 1.95, 111.0, 0.0125, 1.27, 108.0, 0....</td>\n",
       "      <td>0</td>\n",
       "      <td>[0.006609244186046515, 3.2164244186046504, 109...</td>\n",
       "      <td>[-0.8399181449731496, -0.5851533870472007, -0....</td>\n",
       "      <td>AAACT</td>\n",
       "      <td>[0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>316</td>\n",
       "      <td>GAAACAG</td>\n",
       "      <td>[[0.00432, 2.02, 104.0, 0.00299, 3.56, 99.3, 0...</td>\n",
       "      <td>0</td>\n",
       "      <td>[0.0075699999999999995, 2.94054054054054, 105....</td>\n",
       "      <td>[-0.30231932947163226, -0.7279854836073506, -0...</td>\n",
       "      <td>AAACA</td>\n",
       "      <td>[1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_id    transcript_id  position sequence  \\\n",
       "0  ENSG00000004059  ENST00000000233       244  AAGACCA   \n",
       "1  ENSG00000004059  ENST00000000233       261  CAAACTG   \n",
       "2  ENSG00000004059  ENST00000000233       316  GAAACAG   \n",
       "\n",
       "                                               reads  label  \\\n",
       "0  [[0.00299, 2.06, 125.0, 0.0177, 10.4, 122.0, 0...      0   \n",
       "1  [[0.0126, 1.95, 111.0, 0.0125, 1.27, 108.0, 0....      0   \n",
       "2  [[0.00432, 2.02, 104.0, 0.00299, 3.56, 99.3, 0...      0   \n",
       "\n",
       "                                          mean_reads  \\\n",
       "0  [0.008264378378378385, 4.223783783783786, 123....   \n",
       "1  [0.006609244186046515, 3.2164244186046504, 109...   \n",
       "2  [0.0075699999999999995, 2.94054054054054, 105....   \n",
       "\n",
       "                                   scaled_mean_reads middle_sequence  \\\n",
       "0  [0.08622580166837883, -0.06361783691765782, 1....           AGACC   \n",
       "1  [-0.8399181449731496, -0.5851533870472007, -0....           AAACT   \n",
       "2  [-0.30231932947163226, -0.7279854836073506, -0...           AAACA   \n",
       "\n",
       "                                 middle_sequence_OHE  \n",
       "0  [0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, ...  \n",
       "1  [0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...  \n",
       "2  [1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Encode train data first, then use the fitted encoder to transform test data\n",
    "encoder, ohe_train_df = one_hot_encode_DRACH(scaled_train_df_w_middle_seq)\n",
    "encoder, ohe_test_df = one_hot_encode_DRACH(scaled_test_df_w_middle_seq, encoder=encoder)\n",
    "\n",
    "display(ohe_train_df.head(3))\n",
    "display(ohe_test_df.head(3))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### SMOTE"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check label imbalance (FOR SMOTE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.04520309457830705"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Proportion of positive labels in the training set\n",
    "np.mean(ohe_train_df['label'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### By GENE_ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of gene_id with no postive labels:  1885\n",
      "Proportion of gene_id with no positive labels:  0.6118143459915611\n"
     ]
    }
   ],
   "source": [
    "# Proportion of positive labels by gene_id\n",
    "prop_gene = ohe_train_df.groupby('gene_id')['label'].mean()\n",
    "\n",
    "# Number of zeros\n",
    "num_zeros = len(prop_gene[prop_gene == 0])\n",
    "print('Number of gene_id with no postive labels: ', num_zeros)\n",
    "print('Proportion of gene_id with no positive labels: ', num_zeros / len(prop_gene))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "from imblearn.over_sampling import SMOTE\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "# Get the scaled mean reads and one-hot encoded middle sequence columns as dataframe\n",
    "X_resampled_list = [] \n",
    "y_resampled_list = []\n",
    "gene_id_list = []\n",
    "\n",
    "for g_id, group in ohe_train_df.groupby('gene_id'):\n",
    "    X = np.vstack(group['scaled_mean_reads'].values)\n",
    "    y = group['label'].values\n",
    "    \n",
    "    if sum(y) < 10 or sum(y - 1) < 10:  # Skip resampling if the number of positive or negative labels is less than 10\n",
    "        X_resampled_list.append(X)\n",
    "        y_resampled_list.append(y)\n",
    "        gene_id_list.extend([g_id] * len(y))  # Use g_id for the gene_id\n",
    "        continue\n",
    "\n",
    "    # Apply SMOTE to the scaled mean reads and labels\n",
    "    smote = SMOTE(random_state=4262)\n",
    "    X_resampled, y_resampled = smote.fit_resample(X, y)\n",
    "\n",
    "    # Append resampled data to lists\n",
    "    X_resampled_list.append(X_resampled)\n",
    "    y_resampled_list.append(y_resampled)\n",
    "    gene_id_list.extend([g_id] * len(y_resampled))  # Use g_id for the gene_id\n",
    "\n",
    "# Combine all resampled data into final arrays\n",
    "X_resampled_gene = np.vstack(X_resampled_list)\n",
    "y_resampled_gene = np.concatenate(y_resampled_list)\n",
    "\n",
    "# Combine into a final DataFrame if needed\n",
    "resampled_gene_df = pd.DataFrame({\n",
    "    'scaled_mean_reads': list(X_resampled_gene),\n",
    "    'gene_id': gene_id_list,\n",
    "    'label': y_resampled_gene\n",
    "})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Proportion of positive labels in the resampled data: 0.04520309457830705\n",
      "Number of gene_id with no postive labels:  1885\n",
      "Proportion of gene_id with no positive labels:  0.6118143459915611\n"
     ]
    }
   ],
   "source": [
    "# Proportion of positive labels in the resampled data\n",
    "print('Proportion of positive labels in the resampled data:', np.mean(resampled_gene_df['label']))\n",
    "\n",
    "# Proportion of positive labels by gene_id\n",
    "prop_gene = resampled_gene_df.groupby('gene_id')['label'].mean()\n",
    "\n",
    "# Number of zeros\n",
    "num_zeros = len(prop_gene[prop_gene == 0])\n",
    "print('Number of gene_id with no postive labels: ', num_zeros)\n",
    "print('Proportion of gene_id with no positive labels: ', num_zeros / len(prop_gene))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### By MIDDLE_SEQUENCE (DRACH)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of DRACH motifs with no postive labels:  0\n",
      "Proportion of DRACH motifs with no positive labels:  0.0\n"
     ]
    }
   ],
   "source": [
    "# Proportion of positive labels by middle_sequence\n",
    "prop_drach = ohe_train_df.groupby('middle_sequence')['label'].mean()\n",
    "\n",
    "# Number of zeros\n",
    "num_zeros = len(prop_drach[prop_drach == 0])\n",
    "print('Number of DRACH motifs with no postive labels: ', num_zeros)\n",
    "print('Proportion of DRACH motifs with no positive labels: ', num_zeros / len(prop_drach))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "middle_sequence\n",
       "TAACA       1\n",
       "TAACC       2\n",
       "AAACC      10\n",
       "AAACA      32\n",
       "AGACC      51\n",
       "TAACT      53\n",
       "TGACA      59\n",
       "TGACC      79\n",
       "GAACC      95\n",
       "AGACA     111\n",
       "GAACA     136\n",
       "AAACT     209\n",
       "TGACT     355\n",
       "GGACC     417\n",
       "AGACT     429\n",
       "GGACA     550\n",
       "GAACT     653\n",
       "GGACT    1146\n",
       "Name: label, dtype: int64"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Number of positive labels by middle_sequence\n",
    "ohe_train_df[ohe_train_df['label'] == 1].groupby('middle_sequence')['label'].count().sort_values()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "middle_sequence\n",
       "AAACC      1\n",
       "AAACA      6\n",
       "AGACC      9\n",
       "TGACA      9\n",
       "TGACC     11\n",
       "TAACT     18\n",
       "GAACC     29\n",
       "AGACA     36\n",
       "AAACT     46\n",
       "GAACA     51\n",
       "TGACT     75\n",
       "AGACT     96\n",
       "GAACT    130\n",
       "GGACC    135\n",
       "GGACA    141\n",
       "GGACT    294\n",
       "Name: label, dtype: int64"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ohe_test_df[ohe_test_df['label'] == 1].groupby('middle_sequence')['label'].count().sort_values()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Apply SMOTE to each group of DRACH motif"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "from imblearn.over_sampling import SMOTE\n",
    "\n",
    "# Get the scaled mean reads and one-hot encoded middle sequence columns as dataframe\n",
    "X_resampled_list = [] \n",
    "y_resampled_list = []\n",
    "middle_sequence_list = []\n",
    "\n",
    "for middle_seq, group in ohe_train_df.groupby('middle_sequence'):\n",
    "    X = np.vstack(group['scaled_mean_reads'].values)\n",
    "    y = group['label'].values\n",
    "    \n",
    "    if sum(y) < 10: # Skip if the number of positive labels is less than 10\n",
    "        X_resampled_list.append(X)\n",
    "        y_resampled_list.append(y)\n",
    "        middle_sequence_list.extend([middle_seq] * len(y))\n",
    "        continue\n",
    "\n",
    "    # Apply SMOTE to the scaled mean reads and labels\n",
    "    smote = SMOTE(random_state=4262)\n",
    "    X_resampled, y_resampled = smote.fit_resample(X, y)\n",
    "\n",
    "    # Append resampled data and one-hot encoding to lists\n",
    "    X_resampled_list.append(X_resampled)\n",
    "    y_resampled_list.append(y_resampled)\n",
    "    middle_sequence_list.extend([middle_seq] * len(y_resampled))\n",
    "\n",
    "# Combine all resampled data into final arrays\n",
    "X_resampled = np.vstack(X_resampled_list)\n",
    "y_resampled = np.concatenate(y_resampled_list)\n",
    "\n",
    "# Combine into a final DataFrame if needed\n",
    "resampled_drach_df = pd.DataFrame({\n",
    "    'scaled_mean_reads': list(X_resampled),\n",
    "    'middle_sequence': middle_sequence_list,\n",
    "    'label': y_resampled\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Proportion of positive labels in the resampled data:  0.483202765619337\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "middle_sequence\n",
       "AAACA    0.500000\n",
       "AAACC    0.500000\n",
       "AAACT    0.500000\n",
       "AGACA    0.500000\n",
       "AGACC    0.500000\n",
       "AGACT    0.500000\n",
       "GAACA    0.500000\n",
       "GAACC    0.500000\n",
       "GAACT    0.500000\n",
       "GGACA    0.500000\n",
       "GGACC    0.500000\n",
       "GGACT    0.500000\n",
       "TAACA    0.000271\n",
       "TAACC    0.000856\n",
       "TAACT    0.500000\n",
       "TGACA    0.500000\n",
       "TGACC    0.500000\n",
       "TGACT    0.500000\n",
       "Name: label, dtype: float64"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Proportion of positive labels in the resampled data\n",
    "print('Proportion of positive labels in the resampled data: ', np.mean(resampled_drach_df['label']))\n",
    "\n",
    "# Proportion of positive labels by middle_sequence\n",
    "prop_drach_resampled = resampled_drach_df.groupby('middle_sequence')['label'].mean()\n",
    "prop_drach_resampled"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "One hot encode DRACH motifs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>scaled_mean_reads</th>\n",
       "      <th>middle_sequence</th>\n",
       "      <th>label</th>\n",
       "      <th>middle_sequence_OHE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>[2.5249757619759317, -0.7835678231993425, -0.2...</td>\n",
       "      <td>AAACA</td>\n",
       "      <td>0</td>\n",
       "      <td>[1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>[0.6928199968356539, -1.1676038905620205, -0.7...</td>\n",
       "      <td>AAACA</td>\n",
       "      <td>0</td>\n",
       "      <td>[1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>[0.3732826597732115, -0.8372266235169966, -0.2...</td>\n",
       "      <td>AAACA</td>\n",
       "      <td>0</td>\n",
       "      <td>[1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   scaled_mean_reads middle_sequence  label  \\\n",
       "0  [2.5249757619759317, -0.7835678231993425, -0.2...           AAACA      0   \n",
       "1  [0.6928199968356539, -1.1676038905620205, -0.7...           AAACA      0   \n",
       "2  [0.3732826597732115, -0.8372266235169966, -0.2...           AAACA      0   \n",
       "\n",
       "                                 middle_sequence_OHE  \n",
       "0  [1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...  \n",
       "1  [1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...  \n",
       "2  [1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "encoder, resampled_drach_df = one_hot_encode_DRACH(resampled_drach_df, encoder)\n",
    "resampled_drach_df.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Prepare data for model input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# SMOTE by middle_sequence\n",
    "X_resampled_drach = prepare_for_model(resampled_drach_df)\n",
    "y_resampled_drach = resampled_drach_df['label'].values\n",
    "\n",
    "# Without SMOTE\n",
    "X_train = prepare_for_model(ohe_train_df)\n",
    "y_train = ohe_train_df['label'].values\n",
    "\n",
    "# Prepare test data\n",
    "X_test = prepare_for_model(ohe_test_df)\n",
    "y_test = ohe_test_df['label'].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "X_resampled_drach shape: (179345, 27)\n",
      "y_resampled_drach shape: (179345,)\n",
      "X_train shape: (97073, 27)\n",
      "y_train shape: (97073,)\n",
      "X_test shape: (24765, 27)\n",
      "y_test shape: (24765,)\n"
     ]
    }
   ],
   "source": [
    "# print the shapes\n",
    "print('X_resampled_drach shape:', X_resampled_drach.shape)\n",
    "print('y_resampled_drach shape:', y_resampled_drach.shape)\n",
    "print('X_train shape:', X_train.shape)\n",
    "print('y_train shape:', y_train.shape)\n",
    "print('X_test shape:', X_test.shape)\n",
    "print('y_test shape:', y_test.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### NN model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "from tensorflow.keras.models import Sequential\n",
    "from tensorflow.keras.layers import Input, Dense, Dropout\n",
    "from tensorflow.keras.callbacks import ModelCheckpoint\n",
    "from tensorflow.keras.metrics import AUC\n",
    "\n",
    "def build_model(input_shape):\n",
    "    \"\"\"\n",
    "    Define and compile neural network model.\n",
    "    \"\"\"\n",
    "    model = Sequential([\n",
    "        Input(shape=(input_shape,)),\n",
    "        Dense(150, activation='relu'),\n",
    "        Dropout(0.2),  # Dropout layer for regularization\n",
    "        Dense(32, activation='relu'),\n",
    "        Dropout(0.2),  # Another dropout layer\n",
    "        Dense(1, activation='sigmoid')\n",
    "    ])\n",
    "    # Keep track of AUC-PR and AUC-ROC during training\n",
    "    model.compile(\n",
    "        optimizer='adam',\n",
    "        loss='binary_crossentropy',\n",
    "        metrics=[\n",
    "            AUC(curve='PR', name='auc_pr'),  # AUC-PR\n",
    "            AUC(curve='ROC', name='auc_roc')  # AUC-ROC\n",
    "        ]\n",
    "    )\n",
    "    return model"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Train model with and without SMOTE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.model_selection import KFold\n",
    "from tensorflow.keras.callbacks import ModelCheckpoint\n",
    "\n",
    "# Set up 5-fold cross-validation\n",
    "kf = KFold(n_splits=5, shuffle=True, random_state=4262)\n",
    "best_val_auc_pr_smote = 0\n",
    "best_val_auc_pr_no_smote = 0\n",
    "BEST_MODEL_PATH_SMOTE = \"../artifacts/best_model_with_smote.keras\"\n",
    "BEST_MODEL_PATH_NO_SMOTE = \"../artifacts/best_model_without_smote.keras\"\n",
    "\n",
    "# Set checkpoint to save the best model (based on validation AUC-PR)\n",
    "checkpoint_smote = ModelCheckpoint(\n",
    "        BEST_MODEL_PATH_SMOTE,\n",
    "        save_best_only=True,\n",
    "        monitor='val_auc_pr',\n",
    "        mode='max'\n",
    "    )\n",
    "\n",
    "checkpoint_no_smote = ModelCheckpoint(\n",
    "        BEST_MODEL_PATH_NO_SMOTE,\n",
    "        save_best_only=True,\n",
    "        monitor='val_auc_pr',\n",
    "        mode='max'\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "SMOTE model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Training fold 1 with SMOTE\n",
      "Epoch 1/5\n",
      "\u001b[1m4484/4484\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m11s\u001b[0m 2ms/step - auc_pr: 0.6921 - auc_roc: 0.7192 - loss: 0.6040 - val_auc_pr: 0.8304 - val_auc_roc: 0.8472 - val_loss: 0.4916\n",
      "Epoch 2/5\n",
      "\u001b[1m4484/4484\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m9s\u001b[0m 2ms/step - auc_pr: 0.8160 - auc_roc: 0.8352 - loss: 0.4980 - val_auc_pr: 0.8637 - val_auc_roc: 0.8792 - val_loss: 0.4421\n",
      "Epoch 3/5\n",
      "\u001b[1m4484/4484\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m9s\u001b[0m 2ms/step - auc_pr: 0.8435 - auc_roc: 0.8610 - loss: 0.4627 - val_auc_pr: 0.8776 - val_auc_roc: 0.8915 - val_loss: 0.4206\n",
      "Epoch 4/5\n",
      "\u001b[1m4484/4484\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m9s\u001b[0m 2ms/step - auc_pr: 0.8569 - auc_roc: 0.8730 - loss: 0.4440 - val_auc_pr: 0.8880 - val_auc_roc: 0.9014 - val_loss: 0.4015\n",
      "Epoch 5/5\n",
      "\u001b[1m4484/4484\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m10s\u001b[0m 2ms/step - auc_pr: 0.8651 - auc_roc: 0.8813 - loss: 0.4308 - val_auc_pr: 0.8902 - val_auc_roc: 0.9032 - val_loss: 0.3957\n",
      "Training fold 2 with SMOTE\n",
      "Epoch 1/5\n",
      "\u001b[1m4484/4484\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m10s\u001b[0m 2ms/step - auc_pr: 0.6961 - auc_roc: 0.7226 - loss: 0.6014 - val_auc_pr: 0.8364 - val_auc_roc: 0.8522 - val_loss: 0.4783\n",
      "Epoch 2/5\n",
      "\u001b[1m4484/4484\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m9s\u001b[0m 2ms/step - auc_pr: 0.8226 - auc_roc: 0.8420 - loss: 0.4890 - val_auc_pr: 0.8671 - val_auc_roc: 0.8820 - val_loss: 0.4362\n",
      "Epoch 3/5\n",
      "\u001b[1m4484/4484\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m8s\u001b[0m 2ms/step - auc_pr: 0.8497 - auc_roc: 0.8662 - loss: 0.4545 - val_auc_pr: 0.8798 - val_auc_roc: 0.8928 - val_loss: 0.4154\n",
      "Epoch 4/5\n",
      "\u001b[1m4484/4484\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m8s\u001b[0m 2ms/step - auc_pr: 0.8601 - auc_roc: 0.8774 - loss: 0.4370 - val_auc_pr: 0.8925 - val_auc_roc: 0.9046 - val_loss: 0.3970\n",
      "Epoch 5/5\n",
      "\u001b[1m4484/4484\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m8s\u001b[0m 2ms/step - auc_pr: 0.8683 - auc_roc: 0.8833 - loss: 0.4268 - val_auc_pr: 0.8959 - val_auc_roc: 0.9083 - val_loss: 0.3869\n",
      "Training fold 3 with SMOTE\n",
      "Epoch 1/5\n",
      "\u001b[1m4484/4484\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m10s\u001b[0m 2ms/step - auc_pr: 0.6925 - auc_roc: 0.7211 - loss: 0.6017 - val_auc_pr: 0.8333 - val_auc_roc: 0.8543 - val_loss: 0.4784\n",
      "Epoch 2/5\n",
      "\u001b[1m4484/4484\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m8s\u001b[0m 2ms/step - auc_pr: 0.8170 - auc_roc: 0.8358 - loss: 0.4964 - val_auc_pr: 0.8644 - val_auc_roc: 0.8837 - val_loss: 0.4306\n",
      "Epoch 3/5\n",
      "\u001b[1m4484/4484\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m23s\u001b[0m 5ms/step - auc_pr: 0.8447 - auc_roc: 0.8624 - loss: 0.4601 - val_auc_pr: 0.8768 - val_auc_roc: 0.8947 - val_loss: 0.4150\n",
      "Epoch 4/5\n",
      "\u001b[1m4484/4484\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m7s\u001b[0m 2ms/step - auc_pr: 0.8601 - auc_roc: 0.8746 - loss: 0.4408 - val_auc_pr: 0.8864 - val_auc_roc: 0.9023 - val_loss: 0.3987\n",
      "Epoch 5/5\n",
      "\u001b[1m4484/4484\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m8s\u001b[0m 2ms/step - auc_pr: 0.8679 - auc_roc: 0.8826 - loss: 0.4286 - val_auc_pr: 0.8967 - val_auc_roc: 0.9123 - val_loss: 0.3862\n",
      "Training fold 4 with SMOTE\n",
      "Epoch 1/5\n",
      "\u001b[1m4484/4484\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m9s\u001b[0m 2ms/step - auc_pr: 0.6861 - auc_roc: 0.7167 - loss: 0.6044 - val_auc_pr: 0.8419 - val_auc_roc: 0.8569 - val_loss: 0.4772\n",
      "Epoch 2/5\n",
      "\u001b[1m4484/4484\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m8s\u001b[0m 2ms/step - auc_pr: 0.8254 - auc_roc: 0.8414 - loss: 0.4891 - val_auc_pr: 0.8747 - val_auc_roc: 0.8875 - val_loss: 0.4265\n",
      "Epoch 3/5\n",
      "\u001b[1m4484/4484\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m9s\u001b[0m 2ms/step - auc_pr: 0.8510 - auc_roc: 0.8663 - loss: 0.4535 - val_auc_pr: 0.8873 - val_auc_roc: 0.8994 - val_loss: 0.4101\n",
      "Epoch 4/5\n",
      "\u001b[1m4484/4484\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m8s\u001b[0m 2ms/step - auc_pr: 0.8625 - auc_roc: 0.8770 - loss: 0.4359 - val_auc_pr: 0.8953 - val_auc_roc: 0.9080 - val_loss: 0.3920\n",
      "Epoch 5/5\n",
      "\u001b[1m4484/4484\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m8s\u001b[0m 2ms/step - auc_pr: 0.8704 - auc_roc: 0.8849 - loss: 0.4236 - val_auc_pr: 0.8992 - val_auc_roc: 0.9111 - val_loss: 0.3818\n",
      "Training fold 5 with SMOTE\n",
      "Epoch 1/5\n",
      "\u001b[1m4484/4484\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m10s\u001b[0m 2ms/step - auc_pr: 0.6966 - auc_roc: 0.7239 - loss: 0.6001 - val_auc_pr: 0.8394 - val_auc_roc: 0.8536 - val_loss: 0.4806\n",
      "Epoch 2/5\n",
      "\u001b[1m4484/4484\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m7s\u001b[0m 2ms/step - auc_pr: 0.8186 - auc_roc: 0.8364 - loss: 0.4949 - val_auc_pr: 0.8745 - val_auc_roc: 0.8859 - val_loss: 0.4319\n",
      "Epoch 3/5\n",
      "\u001b[1m4484/4484\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m7s\u001b[0m 2ms/step - auc_pr: 0.8474 - auc_roc: 0.8635 - loss: 0.4579 - val_auc_pr: 0.8798 - val_auc_roc: 0.8913 - val_loss: 0.4160\n",
      "Epoch 4/5\n",
      "\u001b[1m4484/4484\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m8s\u001b[0m 2ms/step - auc_pr: 0.8585 - auc_roc: 0.8749 - loss: 0.4395 - val_auc_pr: 0.8915 - val_auc_roc: 0.9039 - val_loss: 0.3988\n",
      "Epoch 5/5\n",
      "\u001b[1m4484/4484\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m8s\u001b[0m 2ms/step - auc_pr: 0.8663 - auc_roc: 0.8830 - loss: 0.4264 - val_auc_pr: 0.8968 - val_auc_roc: 0.9088 - val_loss: 0.3858\n"
     ]
    }
   ],
   "source": [
    "# Cross-validation with SMOTE\n",
    "for fold, (train_idx, val_idx) in enumerate(kf.split(X_resampled_drach)):\n",
    "    print(f\"Training fold {fold + 1} with SMOTE\")\n",
    "\n",
    "    # Split the data with SMOTE for this fold\n",
    "    X_train_smote, X_val_smote = X_resampled_drach[train_idx].copy(), X_resampled_drach[val_idx].copy()\n",
    "    y_train_smote, y_val_smote = y_resampled_drach[train_idx].copy(), y_resampled_drach[val_idx].copy()\n",
    "\n",
    "    # Initialize and compile the model\n",
    "    model_smote = build_model(X_train_smote.shape[1])\n",
    "\n",
    "    # Train the model on this fold with SMOTE data\n",
    "    history_smote = model_smote.fit(\n",
    "        X_train_smote, y_train_smote,\n",
    "        epochs=5,\n",
    "        batch_size=32,\n",
    "        validation_data=(X_val_smote, y_val_smote),\n",
    "        callbacks=[checkpoint_smote]\n",
    "    )\n",
    "\n",
    "    # Track the best validation AUC-PR across folds with SMOTE\n",
    "    fold_best_auc_pr_smote = max(history_smote.history['val_auc_pr'])\n",
    "    if fold_best_auc_pr_smote > best_val_auc_pr_smote:\n",
    "        best_val_auc_pr_smote = fold_best_auc_pr_smote"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "No SMOTE Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Training fold 1 without SMOTE\n",
      "Epoch 1/5\n",
      "\u001b[1m2427/2427\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m6s\u001b[0m 2ms/step - auc_pr: 0.1585 - auc_roc: 0.7597 - loss: 0.1709 - val_auc_pr: 0.3764 - val_auc_roc: 0.8639 - val_loss: 0.1335\n",
      "Epoch 2/5\n",
      "\u001b[1m2427/2427\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m4s\u001b[0m 2ms/step - auc_pr: 0.3460 - auc_roc: 0.8465 - loss: 0.1373 - val_auc_pr: 0.3848 - val_auc_roc: 0.8664 - val_loss: 0.1333\n",
      "Epoch 3/5\n",
      "\u001b[1m2427/2427\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m4s\u001b[0m 2ms/step - auc_pr: 0.3641 - auc_roc: 0.8636 - loss: 0.1330 - val_auc_pr: 0.4083 - val_auc_roc: 0.8768 - val_loss: 0.1265\n",
      "Epoch 4/5\n",
      "\u001b[1m2427/2427\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m4s\u001b[0m 2ms/step - auc_pr: 0.3907 - auc_roc: 0.8594 - loss: 0.1335 - val_auc_pr: 0.4165 - val_auc_roc: 0.8817 - val_loss: 0.1254\n",
      "Epoch 5/5\n",
      "\u001b[1m2427/2427\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m4s\u001b[0m 2ms/step - auc_pr: 0.3900 - auc_roc: 0.8698 - loss: 0.1288 - val_auc_pr: 0.4218 - val_auc_roc: 0.8842 - val_loss: 0.1237\n",
      "Training fold 2 without SMOTE\n",
      "Epoch 1/5\n",
      "\u001b[1m2427/2427\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m6s\u001b[0m 2ms/step - auc_pr: 0.1501 - auc_roc: 0.7607 - loss: 0.1775 - val_auc_pr: 0.3586 - val_auc_roc: 0.8582 - val_loss: 0.1354\n",
      "Epoch 2/5\n",
      "\u001b[1m2427/2427\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m4s\u001b[0m 2ms/step - auc_pr: 0.3485 - auc_roc: 0.8515 - loss: 0.1385 - val_auc_pr: 0.3905 - val_auc_roc: 0.8701 - val_loss: 0.1307\n",
      "Epoch 3/5\n",
      "\u001b[1m2427/2427\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m5s\u001b[0m 2ms/step - auc_pr: 0.3672 - auc_roc: 0.8667 - loss: 0.1320 - val_auc_pr: 0.4065 - val_auc_roc: 0.8753 - val_loss: 0.1280\n",
      "Epoch 4/5\n",
      "\u001b[1m2427/2427\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m5s\u001b[0m 2ms/step - auc_pr: 0.3916 - auc_roc: 0.8683 - loss: 0.1287 - val_auc_pr: 0.4125 - val_auc_roc: 0.8782 - val_loss: 0.1292\n",
      "Epoch 5/5\n",
      "\u001b[1m2427/2427\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m5s\u001b[0m 2ms/step - auc_pr: 0.4077 - auc_roc: 0.8783 - loss: 0.1275 - val_auc_pr: 0.4252 - val_auc_roc: 0.8798 - val_loss: 0.1257\n",
      "Training fold 3 without SMOTE\n",
      "Epoch 1/5\n",
      "\u001b[1m2427/2427\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m7s\u001b[0m 2ms/step - auc_pr: 0.1583 - auc_roc: 0.7537 - loss: 0.1712 - val_auc_pr: 0.3445 - val_auc_roc: 0.8667 - val_loss: 0.1305\n",
      "Epoch 2/5\n",
      "\u001b[1m2427/2427\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m5s\u001b[0m 2ms/step - auc_pr: 0.3520 - auc_roc: 0.8469 - loss: 0.1409 - val_auc_pr: 0.3734 - val_auc_roc: 0.8752 - val_loss: 0.1265\n",
      "Epoch 3/5\n",
      "\u001b[1m2427/2427\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m5s\u001b[0m 2ms/step - auc_pr: 0.3798 - auc_roc: 0.8561 - loss: 0.1360 - val_auc_pr: 0.3936 - val_auc_roc: 0.8724 - val_loss: 0.1254\n",
      "Epoch 4/5\n",
      "\u001b[1m2427/2427\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m5s\u001b[0m 2ms/step - auc_pr: 0.3982 - auc_roc: 0.8664 - loss: 0.1338 - val_auc_pr: 0.4062 - val_auc_roc: 0.8776 - val_loss: 0.1235\n",
      "Epoch 5/5\n",
      "\u001b[1m2427/2427\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m4s\u001b[0m 2ms/step - auc_pr: 0.4110 - auc_roc: 0.8756 - loss: 0.1287 - val_auc_pr: 0.4182 - val_auc_roc: 0.8859 - val_loss: 0.1210\n",
      "Training fold 4 without SMOTE\n",
      "Epoch 1/5\n",
      "\u001b[1m2427/2427\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m6s\u001b[0m 2ms/step - auc_pr: 0.1304 - auc_roc: 0.7460 - loss: 0.1846 - val_auc_pr: 0.3583 - val_auc_roc: 0.8720 - val_loss: 0.1292\n",
      "Epoch 2/5\n",
      "\u001b[1m2427/2427\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m5s\u001b[0m 2ms/step - auc_pr: 0.3496 - auc_roc: 0.8427 - loss: 0.1398 - val_auc_pr: 0.3702 - val_auc_roc: 0.8807 - val_loss: 0.1261\n",
      "Epoch 3/5\n",
      "\u001b[1m2427/2427\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m4s\u001b[0m 2ms/step - auc_pr: 0.3642 - auc_roc: 0.8517 - loss: 0.1335 - val_auc_pr: 0.3823 - val_auc_roc: 0.8852 - val_loss: 0.1243\n",
      "Epoch 4/5\n",
      "\u001b[1m2427/2427\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m4s\u001b[0m 2ms/step - auc_pr: 0.3831 - auc_roc: 0.8605 - loss: 0.1323 - val_auc_pr: 0.3875 - val_auc_roc: 0.8873 - val_loss: 0.1231\n",
      "Epoch 5/5\n",
      "\u001b[1m2427/2427\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m4s\u001b[0m 2ms/step - auc_pr: 0.3977 - auc_roc: 0.8652 - loss: 0.1301 - val_auc_pr: 0.3927 - val_auc_roc: 0.8912 - val_loss: 0.1219\n",
      "Training fold 5 without SMOTE\n",
      "Epoch 1/5\n",
      "\u001b[1m2427/2427\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m6s\u001b[0m 2ms/step - auc_pr: 0.1270 - auc_roc: 0.7425 - loss: 0.1904 - val_auc_pr: 0.3759 - val_auc_roc: 0.8619 - val_loss: 0.1403\n",
      "Epoch 2/5\n",
      "\u001b[1m2427/2427\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m4s\u001b[0m 2ms/step - auc_pr: 0.3252 - auc_roc: 0.8487 - loss: 0.1370 - val_auc_pr: 0.3994 - val_auc_roc: 0.8702 - val_loss: 0.1358\n",
      "Epoch 3/5\n",
      "\u001b[1m2427/2427\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m4s\u001b[0m 2ms/step - auc_pr: 0.3667 - auc_roc: 0.8599 - loss: 0.1356 - val_auc_pr: 0.4154 - val_auc_roc: 0.8758 - val_loss: 0.1354\n",
      "Epoch 4/5\n",
      "\u001b[1m2427/2427\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m5s\u001b[0m 2ms/step - auc_pr: 0.4024 - auc_roc: 0.8707 - loss: 0.1265 - val_auc_pr: 0.4251 - val_auc_roc: 0.8791 - val_loss: 0.1320\n",
      "Epoch 5/5\n",
      "\u001b[1m2427/2427\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m5s\u001b[0m 2ms/step - auc_pr: 0.3912 - auc_roc: 0.8686 - loss: 0.1279 - val_auc_pr: 0.4434 - val_auc_roc: 0.8812 - val_loss: 0.1308\n"
     ]
    }
   ],
   "source": [
    "# Cross-validation without SMOTE\n",
    "for fold, (train_idx, val_idx) in enumerate(kf.split(X_train)):\n",
    "    print(f\"Training fold {fold + 1} without SMOTE\")\n",
    "\n",
    "    # Split the data without SMOTE for this fold\n",
    "    X_train_no_smote, X_val_no_smote = X_train[train_idx].copy(), X_train[val_idx].copy()\n",
    "    y_train_no_smote, y_val_no_smote = y_train[train_idx].copy(), y_train[val_idx].copy()\n",
    "\n",
    "    # Initialize and compile the model\n",
    "    model_no_smote = build_model(X_train_no_smote.shape[1])\n",
    "\n",
    "    # Train the model on this fold without SMOTE data\n",
    "    history_no_smote = model_no_smote.fit(\n",
    "        X_train_no_smote, y_train_no_smote,\n",
    "        epochs=5,\n",
    "        batch_size=32,\n",
    "        validation_data=(X_val_no_smote, y_val_no_smote),\n",
    "        callbacks=[checkpoint_no_smote]\n",
    "    )\n",
    "\n",
    "    # Track the best validation AUC-PR across folds without SMOTE\n",
    "    fold_best_auc_pr_no_smote = max(history_no_smote.history['val_auc_pr'])\n",
    "    if fold_best_auc_pr_no_smote > best_val_auc_pr_no_smote:\n",
    "        best_val_auc_pr_no_smote = fold_best_auc_pr_no_smote"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get the metrics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[1m774/774\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m0s\u001b[0m 350us/step\n",
      "\u001b[1m774/774\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m0s\u001b[0m 334us/step\n",
      "\u001b[1m774/774\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m0s\u001b[0m 356us/step\n",
      "\u001b[1m774/774\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m0s\u001b[0m 336us/step\n",
      "Metrics Comparison on Test Data:\n",
      "\n",
      "With SMOTE:\n",
      "AUC-ROC: 0.8528\n",
      "AUC-PR: 0.3488\n",
      "Accuracy: 0.9580\n",
      "Average Precision Score: 0.3493\n",
      "\n",
      "Without SMOTE:\n",
      "AUC-ROC: 0.8949\n",
      "AUC-PR: 0.4714\n",
      "Accuracy: 0.9561\n",
      "Average Precision Score: 0.4718\n"
     ]
    }
   ],
   "source": [
    "from sklearn.metrics import roc_auc_score, accuracy_score, precision_recall_curve, auc, average_precision_score\n",
    "from tensorflow.keras.models import load_model\n",
    "\n",
    "# Load the best model for both SMOTE and no-SMOTE versions\n",
    "best_model_smote = load_model(BEST_MODEL_PATH_SMOTE)\n",
    "best_model_no_smote = load_model(BEST_MODEL_PATH_NO_SMOTE)\n",
    "\n",
    "# Generate predictions and probabilities for both models on the test data\n",
    "threshold = 0.9 # Threshold for binary predictions\n",
    "\n",
    "y_pred_smote = (best_model_smote.predict(X_test) > threshold).astype(\"int32\").flatten()  # Binary predictions\n",
    "y_proba_smote = best_model_smote.predict(X_test).flatten()  # Probabilities\n",
    "\n",
    "y_pred_no_smote = (best_model_no_smote.predict(X_test) > threshold).astype(\"int32\").flatten()  # Binary predictions\n",
    "y_proba_no_smote = best_model_no_smote.predict(X_test).flatten()  # Probabilities\n",
    "\n",
    "# Calculate metrics for SMOTE model\n",
    "precision, recall, thresholds = precision_recall_curve(y_test, y_proba_smote)\n",
    "roc_auc_smote = roc_auc_score(y_test, y_proba_smote)\n",
    "pr_auc_smote = auc(recall, precision)\n",
    "accuracy_smote = accuracy_score(y_test, y_pred_smote)\n",
    "average_precision_smote = average_precision_score(y_test, y_proba_smote)\n",
    "\n",
    "# Calculate metrics for No-SMOTE model\n",
    "precision, recall, thresholds = precision_recall_curve(y_test, y_proba_no_smote)\n",
    "roc_auc_no_smote = roc_auc_score(y_test, y_proba_no_smote)\n",
    "pr_auc_no_smote = auc(recall, precision)\n",
    "accuracy_no_smote = accuracy_score(y_test, y_pred_no_smote)\n",
    "average_precision_no_smote = average_precision_score(y_test, y_proba_no_smote)\n",
    "\n",
    "# Print comparison results\n",
    "print(\"Metrics Comparison on Test Data:\")\n",
    "print(\"\\nWith SMOTE:\")\n",
    "print(f\"AUC-ROC: {roc_auc_smote:.4f}\")\n",
    "print(f\"AUC-PR: {pr_auc_smote:.4f}\")\n",
    "print(f\"Accuracy: {accuracy_smote:.4f}\")\n",
    "print(f\"Average Precision Score: {average_precision_smote:.4f}\")\n",
    "\n",
    "print(\"\\nWithout SMOTE:\")\n",
    "print(f\"AUC-ROC: {roc_auc_no_smote:.4f}\")\n",
    "print(f\"AUC-PR: {pr_auc_no_smote:.4f}\")\n",
    "print(f\"Accuracy: {accuracy_no_smote:.4f}\")\n",
    "print(f\"Average Precision Score: {average_precision_no_smote:.4f}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Improving Model "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "SMOTE seems to be overfitting the data: try regularisers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#first layer 64"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[1m774/774\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m1s\u001b[0m 1ms/step\n",
      "\u001b[1m774/774\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m1s\u001b[0m 1ms/step\n",
      "\u001b[1m774/774\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m1s\u001b[0m 1ms/step\n",
      "\u001b[1m774/774\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m1s\u001b[0m 1ms/step\n",
      "Metrics Comparison on Test Data:\n",
      "\n",
      "With SMOTE:\n",
      "AUC-ROC: 0.8274\n",
      "AUC-PR: 0.2948\n",
      "Accuracy: 0.9558\n",
      "Average Precision Score: 0.2953\n",
      "\n",
      "Without SMOTE:\n",
      "AUC-ROC: 0.8915\n",
      "AUC-PR: 0.4606\n",
      "Accuracy: 0.9561\n",
      "Average Precision Score: 0.4611\n"
     ]
    }
   ],
   "source": [
    "from sklearn.metrics import roc_auc_score, accuracy_score, precision_recall_curve, auc, average_precision_score\n",
    "from tensorflow.keras.models import load_model\n",
    "\n",
    "# Load the best model for both SMOTE and no-SMOTE versions\n",
    "best_model_smote = load_model(BEST_MODEL_PATH_SMOTE)\n",
    "best_model_no_smote = load_model(BEST_MODEL_PATH_NO_SMOTE)\n",
    "\n",
    "# Generate predictions and probabilities for both models on the test data\n",
    "threshold = 0.9 # Threshold for binary predictions\n",
    "\n",
    "y_pred_smote = (best_model_smote.predict(X_test) > threshold).astype(\"int32\").flatten()  # Binary predictions\n",
    "y_proba_smote = best_model_smote.predict(X_test).flatten()  # Probabilities\n",
    "\n",
    "y_pred_no_smote = (best_model_no_smote.predict(X_test) > threshold).astype(\"int32\").flatten()  # Binary predictions\n",
    "y_proba_no_smote = best_model_no_smote.predict(X_test).flatten()  # Probabilities\n",
    "\n",
    "# Calculate metrics for SMOTE model\n",
    "precision, recall, thresholds = precision_recall_curve(y_test, y_proba_smote)\n",
    "roc_auc_smote = roc_auc_score(y_test, y_proba_smote)\n",
    "pr_auc_smote = auc(recall, precision)\n",
    "accuracy_smote = accuracy_score(y_test, y_pred_smote)\n",
    "average_precision_smote = average_precision_score(y_test, y_proba_smote)\n",
    "\n",
    "# Calculate metrics for No-SMOTE model\n",
    "precision, recall, thresholds = precision_recall_curve(y_test, y_proba_no_smote)\n",
    "roc_auc_no_smote = roc_auc_score(y_test, y_proba_no_smote)\n",
    "pr_auc_no_smote = auc(recall, precision)\n",
    "accuracy_no_smote = accuracy_score(y_test, y_pred_no_smote)\n",
    "average_precision_no_smote = average_precision_score(y_test, y_proba_no_smote)\n",
    "\n",
    "# Print comparison results\n",
    "print(\"Metrics Comparison on Test Data:\")\n",
    "print(\"\\nWith SMOTE:\")\n",
    "print(f\"AUC-ROC: {roc_auc_smote:.4f}\")\n",
    "print(f\"AUC-PR: {pr_auc_smote:.4f}\")\n",
    "print(f\"Accuracy: {accuracy_smote:.4f}\")\n",
    "print(f\"Average Precision Score: {average_precision_smote:.4f}\")\n",
    "\n",
    "print(\"\\nWithout SMOTE:\")\n",
    "print(f\"AUC-ROC: {roc_auc_no_smote:.4f}\")\n",
    "print(f\"AUC-PR: {pr_auc_no_smote:.4f}\")\n",
    "print(f\"Accuracy: {accuracy_no_smote:.4f}\")\n",
    "print(f\"Average Precision Score: {average_precision_no_smote:.4f}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "WITHOUT K FOLD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch 1/5\n",
      "\u001b[1m2427/2427\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m6s\u001b[0m 2ms/step - auc_pr: 0.1884 - auc_roc: 0.7745 - loss: 0.1664 - val_auc_pr: 0.3641 - val_auc_roc: 0.8716 - val_loss: 0.1319\n",
      "Epoch 2/5\n",
      "\u001b[1m2427/2427\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m5s\u001b[0m 2ms/step - auc_pr: 0.3865 - auc_roc: 0.8702 - loss: 0.1311 - val_auc_pr: 0.3756 - val_auc_roc: 0.8741 - val_loss: 0.1353\n",
      "Epoch 3/5\n",
      "\u001b[1m2427/2427\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m5s\u001b[0m 2ms/step - auc_pr: 0.4067 - auc_roc: 0.8660 - loss: 0.1319 - val_auc_pr: 0.3740 - val_auc_roc: 0.8662 - val_loss: 0.1335\n",
      "Epoch 4/5\n",
      "\u001b[1m2427/2427\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m5s\u001b[0m 2ms/step - auc_pr: 0.4150 - auc_roc: 0.8771 - loss: 0.1260 - val_auc_pr: 0.3904 - val_auc_roc: 0.8755 - val_loss: 0.1290\n",
      "Epoch 5/5\n",
      "\u001b[1m2427/2427\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m5s\u001b[0m 2ms/step - auc_pr: 0.4177 - auc_roc: 0.8829 - loss: 0.1256 - val_auc_pr: 0.4029 - val_auc_roc: 0.8822 - val_loss: 0.1276\n",
      "Epoch 1/5\n",
      "\u001b[1m4484/4484\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m10s\u001b[0m 2ms/step - auc_pr: 0.7414 - auc_roc: 0.7724 - loss: 0.5603 - val_auc_pr: 0.7105 - val_auc_roc: 0.6601 - val_loss: 2.7282\n",
      "Epoch 2/5\n",
      "\u001b[1m4484/4484\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m8s\u001b[0m 2ms/step - auc_pr: 0.8491 - auc_roc: 0.8681 - loss: 0.4495 - val_auc_pr: 0.6921 - val_auc_roc: 0.6236 - val_loss: 4.0585\n",
      "Epoch 3/5\n",
      "\u001b[1m4484/4484\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m8s\u001b[0m 2ms/step - auc_pr: 0.8747 - auc_roc: 0.8930 - loss: 0.4090 - val_auc_pr: 0.6624 - val_auc_roc: 0.5855 - val_loss: 5.3381\n",
      "Epoch 4/5\n",
      "\u001b[1m4484/4484\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m8s\u001b[0m 2ms/step - auc_pr: 0.8859 - auc_roc: 0.9027 - loss: 0.3928 - val_auc_pr: 0.6564 - val_auc_roc: 0.5786 - val_loss: 6.2729\n",
      "Epoch 5/5\n",
      "\u001b[1m4484/4484\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m8s\u001b[0m 2ms/step - auc_pr: 0.8974 - auc_roc: 0.9138 - loss: 0.3697 - val_auc_pr: 0.6528 - val_auc_roc: 0.5752 - val_loss: 6.2026\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<keras.src.callbacks.history.History at 0x21a55216950>"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# without kfold\n",
    "no_smote = build_model(X_train.shape[1])\n",
    "no_smote.fit(X_train, y_train, epochs=5, batch_size=32, validation_split=0.2)\n",
    "\n",
    "with_smote = build_model(X_resampled_drach.shape[1])\n",
    "with_smote.fit(X_resampled_drach, y_resampled_drach, epochs=5, batch_size=32, validation_split=0.2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[1m774/774\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m1s\u001b[0m 1ms/step\n",
      "\u001b[1m774/774\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m1s\u001b[0m 1ms/step\n",
      "\u001b[1m774/774\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m1s\u001b[0m 1ms/step\n",
      "\u001b[1m774/774\u001b[0m \u001b[32m\u001b[0m\u001b[37m\u001b[0m \u001b[1m1s\u001b[0m 1ms/step\n",
      "Metrics Comparison on Test Data:\n",
      "\n",
      "With SMOTE:\n",
      "AUC-ROC: 0.8190\n",
      "AUC-PR: 0.2921\n",
      "Accuracy: 0.9567\n",
      "Average Precision Score: 0.2931\n",
      "\n",
      "Without SMOTE:\n",
      "AUC-ROC: 0.8943\n",
      "AUC-PR: 0.4663\n",
      "Accuracy: 0.9561\n",
      "Average Precision Score: 0.4667\n"
     ]
    }
   ],
   "source": [
    "from sklearn.metrics import roc_auc_score, accuracy_score, precision_recall_curve, auc, average_precision_score\n",
    "\n",
    "# Generate predictions and probabilities for both models on the test data\n",
    "threshold = 0.9 # Threshold for binary predictions\n",
    "\n",
    "y_pred_smote = (with_smote.predict(X_test) > threshold).astype(\"int32\").flatten()  # Binary predictions\n",
    "y_proba_smote = with_smote.predict(X_test).flatten()  # Probabilities\n",
    "\n",
    "y_pred_no_smote = (no_smote.predict(X_test) > threshold).astype(\"int32\").flatten()  # Binary predictions\n",
    "y_proba_no_smote = no_smote.predict(X_test).flatten()  # Probabilities\n",
    "\n",
    "# Calculate metrics for SMOTE model\n",
    "precision, recall, thresholds = precision_recall_curve(y_test, y_proba_smote)\n",
    "roc_auc_smote = roc_auc_score(y_test, y_proba_smote)\n",
    "pr_auc_smote = auc(recall, precision)\n",
    "accuracy_smote = accuracy_score(y_test, y_pred_smote)\n",
    "average_precision_smote = average_precision_score(y_test, y_proba_smote)\n",
    "\n",
    "# Calculate metrics for No-SMOTE model\n",
    "precision, recall, thresholds = precision_recall_curve(y_test, y_proba_no_smote)\n",
    "roc_auc_no_smote = roc_auc_score(y_test, y_proba_no_smote)\n",
    "pr_auc_no_smote = auc(recall, precision)\n",
    "accuracy_no_smote = accuracy_score(y_test, y_pred_no_smote)\n",
    "average_precision_no_smote = average_precision_score(y_test, y_proba_no_smote)\n",
    "\n",
    "# Print comparison results\n",
    "print(\"Metrics Comparison on Test Data:\")\n",
    "print(\"\\nWith SMOTE:\")\n",
    "print(f\"AUC-ROC: {roc_auc_smote:.4f}\")\n",
    "print(f\"AUC-PR: {pr_auc_smote:.4f}\")\n",
    "print(f\"Accuracy: {accuracy_smote:.4f}\")\n",
    "print(f\"Average Precision Score: {average_precision_smote:.4f}\")\n",
    "\n",
    "print(\"\\nWithout SMOTE:\")\n",
    "print(f\"AUC-ROC: {roc_auc_no_smote:.4f}\")\n",
    "print(f\"AUC-PR: {pr_auc_no_smote:.4f}\")\n",
    "print(f\"Accuracy: {accuracy_no_smote:.4f}\")\n",
    "print(f\"Average Precision Score: {average_precision_no_smote:.4f}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
